src/dotnet/MLSearch.Service/Engine/Filters.cs

```diff
@@ -0,0 +1,96 @@
+
+using ActualChat.Chat;
+using ActualChat.Contacts;
+using ActualChat.MLSearch.Documents;
+
+namespace ActualChat.MLSearch.Engine;
+
+public interface IFilters
+{
+    ValueTask<IQueryFilter> Semantic(string text, CancellationToken cancellationToken = default);
+    ValueTask<IQueryFilter> Keyword(string text, CancellationToken cancellationToken = default);
+    ValueTask<IQueryFilter> Chat(Func<ChatSet, ChatSet> setBuilder, CancellationToken cancellationToken = default);
+}
+
+internal sealed class Filters(IContactsBackend contacts) : IFilters
+{
+    public ValueTask<IQueryFilter> Semantic(string text, CancellationToken cancellationToken = default)
+        => ValueTask.FromResult<IQueryFilter>(new SemanticFilter<ChatSlice>(text));
+
+    public ValueTask<IQueryFilter> Keyword(string text, CancellationToken cancellationToken = default)
+        => ValueTask.FromResult<IQueryFilter>(new KeywordFilter<ChatSlice>(text.Split()));
+
+    public async ValueTask<IQueryFilter> Chat(Func<ChatSet, ChatSet> setBuilder, CancellationToken cancellationToken = default)
+    {
+        var chatSet = setBuilder.Invoke(new EmptyChatSet());
+
+        var builder = new ChatFilterBuilder(contacts);
+        while (chatSet is not null) {
+            await chatSet.Apply(builder, cancellationToken).ConfigureAwait(false);
+            chatSet = chatSet.Next;
+        }
+        return builder.ChatFilter;
+    }
+}
+
+internal sealed class ChatFilterBuilder(IContactsBackend contacts)
+{
+    public ChatFilter ChatFilter { get; } = new ChatFilter();
+
+    internal ValueTask IncludePublic(PlaceId? placeId, CancellationToken _)
+    {
+        if (placeId.HasValue) {
+            ChatFilter.PlaceIds.Add(placeId.Value);
+        }
+        else {
+            ChatFilter.IncludePublic = true;
+        }
+        return ValueTask.CompletedTask;
+    }
+
+    internal async ValueTask IncludePrivate(UserId userId, PlaceId? placeId, CancellationToken cancellationToken)
+    {
+        var privateContacts = await contacts.ListIdsForSearch(userId, placeId, cancellationToken).ConfigureAwait(false);
```

Comment:
though it will work, I guess you need only private chats. This method returns both private and public afair.

Reply:
You are right, will fix.